### DISCUSSION TITLE
Solution for Shared Folders

### DISCUSSION THREAD

**Discussion by GeoffreyChen777** [Future-Scholars/paperlib#545] [Development]
@igoogolx 目前，用户呼声最高的两个 feature 是：

#524 Shared Folder
#530 Custom Fields

第二个自定义字段比较简单，只需要修改 paper 在数据库里的结构，增加一个数组，就好了。

第一个共享文件夹比较复杂，我目前调研的方案是：

还是利用 mongodb atlas，然后从程序端创建单独的数据库账户，用作共享。

比如我个人云同步的数据库账户是 private@email.com，当这个用户想给别人共享文件夹的时候，自动创建一个 share@email.com 的数据库账户。然后共享文件夹由这个账户创建。

当分享给别人的时候，我们相当于是在分享这个账户 share@email.com ，别的用户订阅这个共享文件夹的时候，实际上就是在自己的客户端登录这个账户。这样就可以共享文件夹了。同时，因为账户不同，共享账户也无法访问到用户私有的论文等数据。

这样我们大部分代码还可以复用，因为本质上就是登陆了多个云同步账户。

不知道你有没有什么好的方案。

**Comment by igoogolx** [comment]
云同步这块，我了解的也不多。这个方案我觉得没问题，那多个同步账户我们怎么管理呢？让用户自己选择当前要使用的账户吗？

**Reply by GeoffreyChen777** [reply]
我是这样设计的，一个账户，对应一个library。

现在实际上我们只有一个library，就是显示在主界面的library。

<img width="421" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/eeb2e851-1317-4895-b6f6-4d49689a97bb">

当用户想要给别人共享的时候，实际上是创建一个新的账号，我想的是，让用户感知不到这个账户，创建的时候我们从代码层面去创建。比如用户云同步用的邮箱是 `A@email.com`，我们可以自动创建 `share-<libname>.A@email.com`，所有的注册登陆链接等，都是代码层面自动完成。用户得到的就是，在这里多出了一个新的 library

<img width="382" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/d1e7daeb-aa93-4a7d-bbf4-2e795571f31d">

同样，其他用户订阅这个人共享的library的时候，也是等于登陆上这个账号，然后这里多出一个 `shared-lib`

然后他们之间的共享，就等价于在同一个账号下的云同步。

这样所有以前的操作逻辑我们都能复用，只需要给那些services实现多账户就好了。UI 激活的是哪个 library，所有的操作就是在对应账号的数据操作。这样就不仅仅是共享文件夹了，直接做到了共享标签，文件夹，智能过滤器等等。

关于权限控制，理论上共享library的共享者可以设置只读和可编辑。我看了看代码SDK无法操作mongodb atlas的权限设置，因此我想这样变通一下：当用户想要修改数据的时候，我们代码里对比这个library所属的账户是不是本机登陆的私人library账户，即 `share-<libname>.A@email.com` 去掉前缀之后等不等于 `A@email.com`，如果是，那就任意编辑，如果不是，根据这个库共享的时候设置的flag操作。
